### ISSUE TITLE
A folder with canView set to false still appears

### ISSUE THREAD

**Issue by mlewis-everley** [silverstripe/silverstripe-asset-admin#982]
We have a intranet system running on SS 4.3

Some users can only have access to certain files and folders. We have overwritten `File::canView()` via an extension.

If the user cannot view the folder it still appears in the files area (`AssetAdmin`). If you then click on the folder you get the error "An unknown error has occurred".

I would have thought the folder should not appear at all?

**Comment by mlewis-everley** [comment]
I had a play around with this a little bit. I found by changing `AssetAdmin.getFiles()` to the following forces asset admin to respect the view state, but the pagination does not...

```
  getFiles() {
    const {
      files,
      queuedFiles,
    } = this.props;

    const combinedFilesList = [
      // Exclude uploaded files that have been reloaded via graphql
      ...queuedFiles
        .items
        .filter(item => !item.id || !files.find(file => file.id === item.id)),
      ...files,
    ];

    const cleanFiles = [];

    // Ensure only viewable files appear
    for (var key in combinedFilesList) {
      if (combinedFilesList[key].canView !== false) {
        cleanFiles.push(combinedFilesList[key]);
      }
    }

    // Seperate folder and files then return an array with folders at the top (for table view)
    const foldersList = cleanFiles.filter((file) => file.type === 'folder');
    const filesList = cleanFiles.filter((file) => file.type !== 'folder');

    return foldersList.concat(filesList);
  }
```

Does anyone have any ideas how the pagination gets a lit of files? If I can work that out I am happy to submit a PR.

**Comment by mlewis-everley** [comment]
@robbieaverill Hate to name drop you on this, wondering if you had any ideas on this issue? I am kind of banging my head on the desk with it...

**Comment by robbieaverill** [comment]
IMO the files/folders shouldn't be returned at all by the GraphQL query when canView() is false

**Comment by mlewis-everley** [comment]
@robbieaverill Makes sense, is this an issue with `silverstripe-graphql`? Or rather how it is implemented in this module?

**Comment by robbieaverill** [comment]
If the GraphQL resolvers are defined in this module (they probably are) then it's an issue for this module. Checking permissions on each record in PHP is costly, however, so it probably won't be an easy fix.

**Comment by mlewis-everley** [comment]
Hmm, yes, I have seen the permissions argument on another issue related to GridField lists. In this case though, the GraphQL response contains the `canView` for each record, so I am guessing this module is checking the permissions anyway?

**Comment by mlewis-everley** [comment]
@robbieaverill I did a bit more digging, it looks like the GraphQL resolvers _should_ already be checking canView permissions (https://github.com/silverstripe/silverstripe-asset-admin/blob/1/code/GraphQL/ReadFileQueryCreator.php#L92) but it appears this it not working as expected. Not really sure what is going on here, this check is pretty simple, files that cannot be viewed should be removed...

**Comment by mlewis-everley** [comment]
Right, I have finally (I think) tracked this issue to `SilverStripe\AssetAdmin\GraphQL\FolderTypeCreator::resolveChildrenConnection()`. Making the following change fixes the issue, but the performance is noticeably effected (even on a few hundred records)

```
    public function resolveChildrenConnection(
        $object,
        array $args,
        $context,
        ResolveInfo $info,
        Connection $childrenConnection
    ) {
        // canView() checks on parent folder are implied by the query returning $object
        // Note: The inability to query permissions against the entire set means pagination
        // is inaccurate when any item in the list returns false on canView()

        $filter = (!empty($args['filter'])) ? $args['filter'] : [];

        if (isset($filter['parentId']) && (int)$filter['parentId'] !== (int)$object->ID) {
            throw new \InvalidArgumentException(sprintf(
                'The "parentId" value (#%d) needs to match the current object id (#%d)',
                (int)$filter['parentId'],
                (int)$object->ID
            ));
        }

        /** @var DataList $list */
        $list = Versioned::get_by_stage(File::class, 'Stage');
        $filterInputType = new FileFilterInputTypeCreator($this->manager);
        
        // Permission check the list
        $list = $list->filterByCallback(function (File $file) use ($context) {
            return $file->canView($context['currentUser']);
        });

        $filter['parentId'] = $object->ID;
        $list = $filterInputType->filterList($list, $filter);

        // Ensure that we're looking at a subset of relevant data.
        $result = $childrenConnection->resolveList($list, $args);
        $list = $result['edges'];

        $result['edges'] = $list;
        $result['pageInfo']['totalCount'] = $list->count();

        return $result;
    }
```

I am not 100% sure how to improve this performance at the moment, it is kind of a pain in the ass if I am honest...

**Comment by robbieaverill** [comment]
cc @Cheddam @emteknetnz, might be related to the permission indicator icon you're adding

**Comment by emteknetnz** [comment]
@robbieaverill this is unrelated the [permission indicator work](https://github.com/silverstripe/silverstripe-assets/pull/398/files#diff-41b5ea2bbaadc935cf6b51194eba6e68R512).  There were do not do a canView() check, instead we only check that permissions exists on a DataObject - we do not check against the current member

**Comment by emteknetnz** [comment]
Without digging to far into this, I have a hunch what's going on here is that `File::canView()` uses `InheritedPermission::canView()` which is much more performant way of doing a `canView()` check on nested DataObjects, however, it does not run any custom php `canView()` methods.

This would explain the outcome originally seen i.e. `We have overwritten File::canView() via an extension.` i.e. the `canView()` extension method is not being called

The proposed code sample where the `canView()` is programmatically called solves this issue, though as noted, it results in a large performance drop

I don't have a recommended way forward at this stage
